src/Microsoft.AspNetCore.Routing/Matchers/CandidateSet.cs

```diff
@@ -2,56 +2,149 @@
+                    _state2 = new CandidateState(endpoints[2], score: 0);
+                    _state3 = new CandidateState(endpoints[3], score: 0);
+
+                    _additionalCandidates = new CandidateState[endpoints.Length - 4];
+                    for (var i = 4; i < endpoints.Length; i++)
+                    {
+                        _additionalCandidates[i - 4] = new CandidateState(endpoints[i], score: 0);
+                    }
+                    break;
+            }
         }
 
-        // See description on Groups.
-        public static int[] MakeGroups(int[] lengths)
+        internal CandidateSet(Candidate[] candidates)
         {
-            var groups = new int[lengths.Length + 1];
+            Count = candidates.Length;
 
-            var sum = 0;
-            for (var i = 0; i < lengths.Length; i++)
+            switch (candidates.Length)
             {
-                groups[i] = sum;
-                sum += lengths[i];
+                case 0:
+                    return;
+
+                case 1:
+                    _state0 = new CandidateState(candidates[0].Endpoint, candidates[0].Score);
+                    break;
+
+                case 2:
+                    _state0 = new CandidateState(candidates[0].Endpoint, candidates[0].Score);
+                    _state1 = new CandidateState(candidates[1].Endpoint, candidates[1].Score);
+                    break;
+
+                case 3:
+                    _state0 = new CandidateState(candidates[0].Endpoint, candidates[0].Score);
+                    _state1 = new CandidateState(candidates[1].Endpoint, candidates[1].Score);
+                    _state2 = new CandidateState(candidates[2].Endpoint, candidates[2].Score);
+                    break;
+
+                case 4:
+                    _state0 = new CandidateState(candidates[0].Endpoint, candidates[0].Score);
+                    _state1 = new CandidateState(candidates[1].Endpoint, candidates[1].Score);
+                    _state2 = new CandidateState(candidates[2].Endpoint, candidates[2].Score);
+                    _state3 = new CandidateState(candidates[3].Endpoint, candidates[3].Score);
+                    break;
+
+                default:
+                    _state0 = new CandidateState(candidates[0].Endpoint, candidates[0].Score);
+                    _state1 = new CandidateState(candidates[1].Endpoint, candidates[1].Score);
+                    _state2 = new CandidateState(candidates[2].Endpoint, candidates[2].Score);
+                    _state3 = new CandidateState(candidates[3].Endpoint, candidates[3].Score);
+
+                    _additionalCandidates = new CandidateState[candidates.Length - 4];
+                    for (var i = 4; i < candidates.Length; i++)
+                    {
+                        _additionalCandidates[i - 4] = new CandidateState(candidates[i].Endpoint, candidates[i].Score);
+                    }
```

Comment:
It appears like there's a lot of memory copying happening here, but my profiling so far shows this cost to be pretty minimal.